File: src/rl_agent/reward.R
Reward <- function(pair, coef, Action_track, cost,rwrd_fun='raw_return', eta_dsr=0.5){

  '
  Returns the Return for each (s,a) pair in the episode.
  The Return is part of the RL framework and is defined
  as the sum of the rewards in an episode.
  
  Input:
  
    pair - data frame containing price series of both assets
    of length equal to the length of the episode
    
    coef - cointegrating coefficient for the assets in the pair
    
    Action_track - vector of actions carried out in the episode
    
    cost - cost of engaging in a trade (changing action)
  '
  
  # get the number of observations
  n <- NROW(pair)
  
  # initialise vector of returns. The reward is defined as 
  # percentage return
  additive <- numeric(NROW(pair[,1]))
  
  
    # reward in the last visited state interval is 0
    # therefor up to n-1
    for(i in 1:(n-1)) {
      
      # calculate return of the paird following each (s,a) pair in the episode.
      # An episode consists only of a=1 or a=-1 and a final different action which
      # termiantes the episode.
      
      if (Action_track[i] == -1) {      
        additive[i] <- (-(coef[1]*(pair[n,1]-pair[i,1]) + coef[2]*(pair[n,2] - pair[i,2])) / (pair[i,1] + abs(coef[2])*pair[i,2]))
        
      } else if (Action_track[i] == 1) { #Action_track taken (long)      
        additive[i] <- (coef[1]*(pair[n,1]-pair[i,1]) + coef[2]*(pair[n,2] - pair[i,2])) / (pair[i,1] + abs(coef[2])*pair[i,2])     
      }
    }
  
  if(rwrd_fun == "dsr"){
    additive <-c(additive[length(additive)], additive[1:(length(additive) - 1)])
    dsr <- diff_sharpe_ratio(additive, eta = eta_dsr)
    reward <- c(dsr[2:length(dsr)], dsr[1])
  }else{
    reward <- additive
  }
  
  return(reward)  
}

Qreward <- function(pair, coef, Action_track, cost, rwrd_fun = 'raw_return', eta_dsr=0.5){
  
  '
  Returns the Return for each (s,a) pair in the episode.
  The Return is part of the RL framework and is defined
  as the sum of the rewards in an episode.
  
  Input:
  
  pair - data frame containing price series of both assets
  of length equal to the length of the episode
  
  coef - cointegrating coefficient for the assets in the pair
  
  Action_track - vector of actions carried out in the episode
  
  cost - cost of engaging in a trade (changing action)
  '
  
  # get the number of observations
  n <- nrow(pair)
  
  # initialise vector of returns. The reward is defined as 
  # percentage return
  additive <- numeric(NROW(pair[,1]) - 1)
  
  # reward in the last visited state interval is 0
  # therefor up to n-1
  for(i in 1:(n-1)) {
    
    # calculate return of the paird following each (s,a) pair in the episode.
    # An episode consists only of a=1 or a=-1 and a final different action which
    # termiantes the episode.
    if (Action_track[i] == -1) {      
      additive[i] <- (-(coef[1]*(pair[n,1]-pair[i,1]) + coef[2]*(pair[n,2] - pair[i,2])) / (pair[i,1] + abs(coef[2])*pair[i,2]))       
    } else if (Action_track[i] == 1) { #Action_track taken (long)      
      additive[i] <- (coef[1]*(pair[n,1]-pair[i,1]) + coef[2]*(pair[n,2] - pair[i,2])) / (pair[i,1] + abs(coef[2])*pair[i,2])     
    } else if (Action_track[i] == 0) {
      additive[i] <- 0
    }
    
  }
  rwrd <- additive
  
  
  if(rwrd_fun == "dsr") {
    additive <- c(0, additive)
    dsr <- diff_sharpe_ratio(additive, eta = eta_dsr)
    rwrd <- c(dsr[2:length(dsr)])
  }
  
  return(rwrd)  
}

episodeBackup <- function(Action_track, State_track, rwrd, diff, coef, cost, Qsa, alpha, Pre_Ns, Ns, Nsa){
  
  actions <- c(-1,0,1) #short, neutral, long
  
  # Episode terminated with a 0
  if(diff == 1){
                   
    for (l in 1:nrow(State_track))
    {
      action <- Action_track[l]
      state <- as.numeric(State_track[l,])
      
      action_dim <- match(action, actions)
      state_action <- as.numeric(c(state, action_dim))
      
      counts <- UpdateN(state, Ns, Nsa, actions, action, cut_points)
      Ns <- counts[[1]]  
      Nsa <- counts[[2]]
      
      #alpha <- 1/Nsa[matrix(state_action, 1)]
      alpha <- 1/(log(Nsa[matrix(state_action, 1)])+1)
      # MCC backup
      Qsa[matrix(state_action, 1)] <- Qsa[matrix(state_action, 1)] + alpha*(rwrd[l] - Qsa[matrix(state_action, 1)])        
    }
  }
  
  if(diff == 2)
  {   
    for (j in 1:nrow(State_track))
    {
      action <- Action_track[j]
      state <- as.numeric(State_track[j,])
      
      action_dim <- match(action, actions)
      state_action <- as.numeric(c(state, action_dim))
      
      counts <- UpdateN(state, Ns, Nsa, actions, action, cut_points)
      Ns <- counts[[1]]  
      Nsa <- counts[[2]]
      
      #alpha <- 1/Nsa[matrix(state_action, 1)]
      alpha <- 1/(log(Nsa[matrix(state_action, 1)])+1) 
      Qsa[matrix(state_action, 1)] <- Qsa[matrix(state_action, 1)] + alpha*(rwrd[j] - Qsa[matrix(state_action, 1)])
    }
    State_track <- State_track[nrow(State_track),]
    Action_track <- Action_track[length(Action_track)]
    # Epsilon updates need to use the new
    Pre_Ns <- Ns
  }
 
  return(list(Action_track, State_track, Qsa, alpha, Ns, Nsa, Pre_Ns))
}


save <- function(pair, coef, spread, Action_track, cost) {
  
  # standard percentage return   
  rwrd <- Ret(pair, coef, spread, Action_track, cost, FALSE)
  rwrd <- rwrd[[1]]
  rwrd <- rwrd[-c(1)] - 1 # back to percentage return
  
  Ret2(pair, coef, spread, Action_track, cost)
  
  #     #get the number of observations
  #     n <- NROW(spread)
  #     
  #     #initialise vector of returns
  #     additive <- numeric(NROW(pair[,1]))
  #     multiplicative <- numeric(NROW(pair[,1])) + 1 
  #     multiplicative[1] <- 1 - cost #cost for opening Action_track
  #        
  #     for(i in 2:n) {
  #       if (Action_track == -1) { 
  #         multiplicative[i] <- ((-(coef[1]*(pair[i,1]-pair[i-1,1]) + coef[2]*(pair[i,2] - pair[i-1,2])) /
  #                            (pair[i-1,1] + abs(coef[2])*pair[i-1,1])) + 1) * multiplicative[i-1]  
  #         additive[i] <- (-(coef[1]*(pair[n,1]-pair[i-1,1]) + coef[2]*(pair[n,2] - pair[i-1,2])) /
  #                                  (pair[i-1,1] + abs(coef[2])*pair[i-1,1]))  
  #       } else if (Action_track == 1) { #Action_track taken (long)
  #         multiplicative[i] <- ((+(coef[1]*(pair[i,1]-pair[i-1,1]) + coef[2]*(pair[i,2] - pair[i-1,2])) /
  #                            (pair[i-1,1] + abs(coef[2])*pair[i-1,1])) + 1) * multiplicative[i-1] 
  #         additive[i] <- (+(coef[1]*(pair[n,1]-pair[i-1,1]) + coef[2]*(pair[n,2] - pair[i-1,2])) /
  #                            (pair[i-1,1] + abs(coef[2])*pair[i-1,1]))
  #       }
  #     }
  #     
  #     #include the cost
  #     multiplicative[n] <- multiplicative[n] * (1-cost)
  #     additive <- additive - cost*2
  #     
  #     additive is the actual reward vector for an episode of MCC
  #     return(list(additive, multiplicative, multiplicative[n]))
  return(rwrd)
}  

File: src/rl_agent_old/reward.py
"""
Reward functions for reinforcement learning trading agent.
"""

import numpy as np


def reward(
    pair: np.ndarray,
    coef: np.ndarray,
    action_track: np.ndarray,
    cost: float,
    reward_function: str = "raw_return",
    eta_dsr: float = 0.5,
) -> np.ndarray:
    """
    Returns the Return for each (s,a) pair in the episode.
    The Return is part of the RL framework and is defined
    as the sum of the rewards in an episode.

    Args:
        pair: DataFrame containing price series of both assets
              of length equal to the length of the episode
        coef: Cointegrating coefficient for the assets in the pair
        action_track: Vector of actions carried out in the episode
        cost: Cost of engaging in a trade (changing action)
        reward_function: Type of reward function to use
        eta_dsr: Parameter for differential Sharpe ratio

    Returns:
        Array of rewards for each state-action pair
    """
    # Get the number of observations
    n = len(pair)

    # Initialize vector of returns. The reward is defined as
    # percentage return
    additive = np.zeros(len(pair[:, 0]))

    # Reward in the last visited state interval is 0
    # Therefore up to n-1
    for i in range(n - 1):
        # Calculate return of the pair following each (s,a) pair in the episode.
        # An episode consists only of a=1 or a=-1 and a final different action which
        # terminates the episode.

        if action_track[i] == -1:
            additive[i] = -(
                coef[0] * (pair[n - 1, 0] - pair[i, 0])
                + coef[1] * (pair[n - 1, 1] - pair[i, 1])
            ) / (pair[i, 0] + abs(coef[1]) * pair[i, 1])

        elif action_track[i] == 1:  # Action taken (long)
            additive[i] = (
                coef[0] * (pair[n - 1, 0] - pair[i, 0])
                + coef[1] * (pair[n - 1, 1] - pair[i, 1])
            ) / (pair[i, 0] + abs(coef[1]) * pair[i, 1])

    if reward_function == "dsr":
        additive = np.concatenate(([additive[-1]], additive[:-1]))
        dsr = diff_sharpe_ratio(additive, eta=eta_dsr)
        reward_values = np.concatenate((dsr[1:], [dsr[0]]))
    else:
        reward_values = additive

    return reward_values


def diff_sharpe_ratio(returns: np.ndarray, eta: float = 0.5) -> np.ndarray:
    """
    Calculate the differential Sharpe ratio.

    Args:
        returns: Array of returns
        eta: Time scale parameter

    Returns:
        Array of differential Sharpe ratios
    """
    # Initialize A and B
    A = np.zeros_like(returns)
    B = np.zeros_like(returns)

    # Calculate K_eta
    K_eta = (1 - eta / 2) / (1 - eta)

    # Calculate A and B recursively
    for t in range(1, len(returns)):
        A[t] = eta * returns[t] + (1 - eta) * A[t - 1]
        B[t] = eta * returns[t] ** 2 + (1 - eta) * B[t - 1]

    # Calculate Sharpe ratio
    S = A / (K_eta * np.sqrt(B - A**2))

    # Calculate differential Sharpe ratio
    dsr = np.zeros_like(returns)
    dsr[1:] = S[1:] - S[:-1]

    return dsr


def q_reward(
    pair: np.ndarray,
    coef: np.ndarray,
    action: int,
    cost: float,
    reward_func: str = "raw_return",
    eta_dsr: float = 0.5,
) -> float:
    """
    Calculate reward for Q-learning agent.

    Args:
        pair: Price data for the trading pair
        coef: Cointegration coefficients
        action: Action taken (-1: short, 0: neutral, 1: long)
        cost: Transaction cost
        reward_func: Reward function to use ('dsr' or 'raw_return')
        eta_dsr: Time scale parameter for differential Sharpe ratio

    Returns:
        Reward value
    """
    # Calculate raw return
    if action == -1:  # Short position
        raw_return = -(
            coef[0] * (pair[1, 0] - pair[0, 0]) + coef[1] * (pair[1, 1] - pair[0, 1])
        ) / (pair[0, 0] + abs(coef[1]) * pair[0, 1])
    elif action == 1:  # Long position
        raw_return = (
            coef[0] * (pair[1, 0] - pair[0, 0]) + coef[1] * (pair[1, 1] - pair[0, 1])
        ) / (pair[0, 0] + abs(coef[1]) * pair[0, 1])
    else:  # Neutral position
        raw_return = 0

    # Apply reward function
    if reward_func == "dsr":
        # For DSR, we need at least 2 returns
        returns = np.array([0, raw_return])  # Add 0 for t=0
        reward = diff_sharpe_ratio(returns, eta=eta_dsr)[-1]
    else:
        reward = raw_return

    return reward


def episode_backup(
    action_track: list[int],
    state_track: np.ndarray,
    reward: np.ndarray,
    diff: int,
    coef: np.ndarray,
    cost: float,
    qsa: np.ndarray,
    alpha: float,
    pre_ns: np.ndarray,
    ns: np.ndarray,
    nsa: np.ndarray,
) -> tuple:
    """
    Perform episode backup for Monte Carlo Control.

    Args:
        action_track: List of actions taken
        state_track: Array of states visited
        reward: Array of rewards received
        diff: Difference type (1 or 2)
        coef: Cointegration coefficients
        cost: Transaction cost
        qsa: Q-value matrix
        alpha: Learning rate
        pre_ns: Previous state visit counts
        ns: Current state visit counts
        nsa: State-action visit counts

    Returns:
        Tuple of updated parameters
    """
    actions = [-1, 0, 1]  # short, neutral, long

    if diff == 1:  # Episode terminated with a 0
        for l in range(len(state_track)):
            action = action_track[l]
            state = state_track[l]

            action_dim = actions.index(action)
            state_action = np.array([*state, action_dim])

            # Update visit counts
            counts = update_n(state, ns, nsa, actions, action)
            ns, nsa = counts

            # Update learning rate
            alpha = 1 / (np.log(nsa[tuple(state_action)] + 1))

            # MCC backup
            qsa[tuple(state_action)] += alpha * (reward[l] - qsa[tuple(state_action)])

    elif diff == 2:
        for j in range(len(state_track)):
            action = action_track[j]
            state = state_track[j]

            action_dim = actions.index(action)
            state_action = np.array([*state, action_dim])

            # Update visit counts
            counts = update_n(state, ns, nsa, actions, action)
            ns, nsa = counts

            # Update learning rate
            alpha = 1 / (np.log(nsa[tuple(state_action)] + 1))

            # Update Q-values
            qsa[tuple(state_action)] += alpha * (reward[j] - qsa[tuple(state_action)])

        # Update state and action tracks
        state_track = state_track[-1:]
        action_track = action_track[-1:]
        pre_ns = ns

    return action_track, state_track, qsa, alpha, ns, nsa, pre_ns
